node/actors/bft/src/testonly/twins/scenario.rs

```diff
@@ -0,0 +1,171 @@
+
+        let mut nodes = replicas;
+        nodes.extend(twins);
+
+        Self {
+            nodes,
+            num_replicas,
+        }
+    }
+
+    /// The number of nodes in the cluster, both replicas and twins.
+    pub fn num_nodes(&self) -> usize {
+        self.nodes.len()
+    }
+
+    /// The number of original replicas, ie. not twins.
+    ///
+    /// This is the `n` in the `n = 5 * f + 1` equation.
+    pub fn num_replicas(&self) -> usize {
+        self.num_replicas
+    }
+
+    /// The number of twins in the cluster.
+    pub fn num_twins(&self) -> usize {
+        self.num_nodes() - self.num_replicas()
+    }
+
+    /// The number of faulty nodes the committee can tolerate for the `n = 5 * f + 1` relation to hold.
+    pub fn num_faulty(&self) -> usize {
+        (self.num_replicas - 1) / 5
+    }
+
+    /// The number of signatures required for a quorum, ie. `4 * f + 1` or more generally `n - f`, which in this case equals `n*4/5+1`
+    pub fn quorum_size(&self) -> usize {
+        self.num_replicas() - self.num_faulty()
+    }
+
+    /// The number of signatures required for a high-vote, ie. `2 * f + 1`
+    pub fn subqourum_size(&self) -> usize {
+        self.num_faulty() * 2 + 1
+    }
+
+    /// All nodes in the cluster, replicas and twins.
+    pub fn nodes(&self) -> &[T] {
+        &self.nodes
+    }
+
+    /// The original replicas in the cluster.
+    pub fn replicas(&self) -> &[T] {
+        &self.nodes[..self.num_replicas]
+    }
+
+    /// The twins we created in the cluster.
+    pub fn twins(&self) -> &[T] {
+        &self.nodes[self.num_replicas..]
+    }
+}
+
+/// The configuration for a single round.
+pub struct RoundConfig<'a, T: HasKey> {
+    pub leader: &'a T::Key,
+    pub partitions: Split<'a, T>,
+}
+
+/// Configuration for a number of rounds.
+pub struct Scenario<'a, T: HasKey> {
+    pub rounds: Vec<RoundConfig<'a, T>>,
+}
+
+pub struct ScenarioGenerator<'a, T>
+where
+    T: Twin,
+{
+    /// Number of rounds (views) to simulate in a scenario
+    num_rounds: usize,
+    /// Unique leader keys.
+    keys: Vec<&'a T::Key>,
+    /// All splits of various sizes we can choose in a round.
+    splits: Vec<Split<'a, T>>,
+}
+
+impl<'a, T> ScenarioGenerator<'a, T>
+where
+    T: Twin,
+{
+    /// Initialise a scenario generator from a cluster.
+    pub fn new(cluster: &'a Cluster<T>, num_rounds: usize, max_partitions: usize) -> Self {
+        assert!(!cluster.nodes().is_empty(), "empty cluster");
+
+        // Potential leaders
+        let keys = cluster.replicas().iter().map(|r| r.key()).collect();
+
+        // Create all possible partitionings; the paper considers 2 or 3 partitions to be enough.
+        let splits = (1..=max_partitions).flat_map(|np| splits(cluster.nodes(), np));
+
+        // Prune partitionings so that all of them have at least one where a quorum is possible.
+        // Alternatively we could keep all and make sure every scenario has eventually one with a quorum, for liveness.
```

Comment:
One thing to consider here would be how the network is implemented in the tests. My idea was that I'd look at the `view` of the message, and then look at the partitioning in that round, and deliver to all nodes who are in the same partition as the sender. 

The ChonkyBFT spec repeats timeouts periodically, however if ever node would be in a partition with a size less than 4f+1, then they would never have a TimeoutQuorum, and would never move on to the next view. 

Whereas if at least one partition has 4f+1 views, then they can time out, and in the next view they might be broadcasting to their stranded brethren, who at that point can move on to the future view on which they now observe a quorum in the justification.

Reply:
The currently implemented network semantics is that all (consensus) messages are delivered eventually, unless overridden by a newer message (see MessagePool for exact semantics).

Reply:
The best approach IMO is to unpartition every scenario at the end. Partitions are a temporary network condition anyway, with these tests we want to see if the consensus can recover from a partition.